tests/scripts/generate_bignum_tests.py

```diff
@@ -0,0 +1,299 @@
+#!/usr/bin/env python3
+"""Generate test data for bignum functions.
+
+With no arguments, generate all test data. With non-option arguments,
+generate only the specified files.
+"""
+
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import itertools
+import os
+import posixpath
+import re
+import sys
+from typing import Iterable, Iterator, List, Optional, Tuple, TypeVar
+
+import scripts_path # pylint: disable=unused-import
+from mbedtls_dev import build_tree
+from mbedtls_dev import test_case
+
+T = TypeVar('T') #pylint: disable=invalid-name
+
+def hex_to_int(val):
+    return int(val, 16) if val else 0
+
+def quote_str(val):
+    return "\"{}\"".format(val)
+
+
+class BaseTarget:
+    """Base target for test case generation.
+
+    Attributes:
+        count: Counter for test class.
+        desc: Short description of test case.
+        func: Function which the class generates tests for.
+        gen_file: File to write generated tests to.
+        title: Description of the test function/purpose.
+    """
+    count = 0
+    desc = ""
+    func = ""
+    gen_file = ""
+    title = ""
+
+    def __init__(self) -> None:
+        type(self).count += 1
+
+    @property
+    def args(self) -> List[str]:
+        """Create list of arguments for test case."""
+        return []
+
+    @property
+    def description(self) -> str:
+        """Create a numbered test description."""
```

Comment:
I think that having numbered test is useful and is worth the cost of that 3-4 characters in the title. Comparing test results across different commits is only a problem if we added test cases or changed generation order between the two commits. Also, in the overwhelming majority of cases we are looking at the tests within a single commit, where having a number makes identifying the failing test cases much easier.

Could we please keep the test case numbering?

Reply:
There is value in having an easy-to-compare unique identifier on test cases. That unique identifier doesn't need to be meaningful, so it can be a number. Now the question is, what kind of stability do we expect on this number?

At one extreme, we can just use the line number. This changes pretty much whenever we add or remove test cases (or comments). Nice properties: lookup by number is extremely easy; the numbers are increasing (but not consecutive); the numbering can be done fully automatically by the test framework.

At another extreme, we can assign a number (or other short identifier) when we write a test case, and change to it. Advantage: the unique identifier is permanent. Downside: has to be managed manually, may cause conflicts if two concurrent tracks of work add the same unique identifier to different test cases.

With the intermediate solution we'd be introducing here, test case numbers would apply only to automatically generated test cases, so this isn't a general solution. We can easily tweak the scope of the numbering (per test function, per data file…) but any scope has its own downside (per test function: makes it hard to figure out what the unique part is; per test suite: makes the numbers unstable when we add unrelated test cases). I think this is not a good design for test case numbering. So I maintain my opposition.

I think the idea of test numbering is a good thing. I, too, have spent time copy-pasting descriptions around to search them, and sometimes looking at the wrong test case because I missed a difference in the description. So far it's been low on my tech debt list, but seeing I'm not the only one interested, then by all means let's add it to our tech debt. But [let's first agree on a design](https://github.com/Mbed-TLS/mbedtls/issues/6227), and introduce it as its own feature rather than part of some unrelated work.

Reply:
I am happy to have discussion about the design, but could we please not block this PR with that discussion?

This is a test framework, whatever we decide, won't affect the test cases we add while discussing the matter. I really would like to start using this in the Bignum refactoring work as soon as possible. We could tackle the global numbering problem in a different PR.

Reply:
> could we please not block this PR with that discussion?

Agreed, absolutely. Test case numbers are going to cause merge conflicts if we work on this script in parallel, which is likely in the near future. So let's not put them in now, and add them later if we decide on that strategy.

Reply:
Great, thanks, sounds good!

Regarding the temporary presence of test numbering in automated tests:

The script output is generated files, they won't be committed and won't cause merge conflicts. If we work in parallel on the script, we won't change the part that adds the numbering and the other PR will likely remove it. Again, it seems as a clean merge, I can't see the conflict.

On the Bignum refactoring side, while we don't have the global numbering solution, having this local numbering would be really helpful. Some tests have hundreds of test cases, it is very difficult to navigate them without any numbering.

Reply:
> The script output is generated files, they won't be committed and won't cause merge conflicts

Well, I was thinking of 2.2x, where they are committed. But it's true that on 2.2x, we typically have quick cycle from making the backport to merging it, so the risk of conflicts is small.

Reply:
Is this going to be backported? I know it says "Yes" in the description, but this is clearly a new feature. And many of the newly-generated tests won't be applicable to bignum-2.28. Of course, having them the same has advantages, but I thought we'd agreed that bignum-development and bignum-2.28 would diverge?

Reply:
This is a test framework, which is helpful to have on both and the test framework doesn't have to (or shouldn't) diverge. Also, this would make backporting tests easier. (We are adding more tests to other areas as well, not just the new code.)